pravega/pravega
# Issue: 6070 Configuration of TLSv1.3 in Pravega Components

Signed-off-by: kotlasaicharanreddy <kotlasaicharan@yahoo.com>

**Change log description**  

- This PR implements `TLSv1.3` support in Pravega components: `Segment store` and `Controller`.
- Segment store and Controller's GRPC endpoint uses `Netty` library to support secure communication.
- Controller's REST endpoint uses `Grizzly` library for TLS communication.
- This PR also includes changes for pravega standalone to support `TLSv1.3` protocol.
- Implements Strict mode (either TLSv1.2 or TLSv1.3) or mixed mode (TLSv1.2 and TLSv1.3) communication.

Prominent files with changes:

- [GRPCServerConfigImpl](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/controller/src/main/java/io/pravega/controller/server/rpc/grpc/impl/GRPCServerConfigImpl.java)
- [RESTServerConfigImpl](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/shared/rest/src/main/java/io/pravega/shared/rest/impl/RESTServerConfigImpl.java)
- [TLSHelper](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/security/TLSHelper.java)
- [ServiceConfig](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java)
- [Config](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/controller/src/main/java/io/pravega/controller/util/Config.java)
- [TLSProtocolVersion](https://github.com/pravega/pravega/blob/8c94d005009381fdecd0736327877ac5ca67de41/common/src/main/java/io/pravega/common/security/TLSProtocolVersion.java)

**Purpose of the change**  
Fixes #6070 

**What the code does**  
(Detailed description of the code changes)

- A new property `security.tls.protocolVersion` is added to provide TLS property version in Controller and Segment Store properties.
- Valid values are `TLSv1.2` (strict) or `TLSv1.3` (strict) or `TLSv1.2,TLSv1.3` (mixed). 
- The default value is the mixed mode to support the clients which can communicate with server only in TLSv1.2

Segment Store changes

- In [TLSHelper](https://github.com/pravega/pravega/blob/2882fdabaa24227385f54caa3d74e936a29eef70/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/security/TLSHelper.java) class following change is made to facilitate TLS protocol version in segment store server.
```
SslContext result = SslContextBuilder.forServer(certificateFile, serverKeyFile).protocols(tlsProtocolVersion).build();
```
Controller Changes

- Controller changes include both GRPC and REST support for TLSv1.3. 

1. GRPC Changes
    In [GRPCServer](https://github.com/pravega/pravega/blob/2882fdabaa24227385f54caa3d74e936a29eef70/controller/src/main/java/io/pravega/controller/server/rpc/grpc/GRPCServer.java) class `SSLContext `is utilized to support TLS protocol version in Controller's GRPC Server.
```
SslContext ctx = GrpcSslContexts.forServer(new File(serverConfig.getTlsCertFile()), new File(serverConfig.getTlsKeyFile())).protocols(serverConfig.getTlsProtocolVersion()).build();
```

2. REST Changes
    In [RESTServer](https://github.com/pravega/pravega/blob/2882fdabaa24227385f54caa3d74e936a29eef70/shared/rest/src/main/java/io/pravega/shared/rest/RESTServer.java) class `httpServer `is modified to support TLS protocol version in Controller's REST server side.
```
httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true,
                        new SSLEngineConfigurator(contextConfigurator, false, false, false).setEnabledProtocols(restServerConfig.tlsProtocolVersion()));
```


 

**How to verify it**  

- All tests must pass and build should succeed.
- Run [TlsEnabledInProcPravegaClusterTest](https://github.com/pravega/pravega/blob/master/standalone/src/test/java/io/pravega/local/TlsEnabledInProcPravegaClusterTest.java) and inspecting logs will show the TLS protocol version supported by Pravega.
